File: library/src/main/webapp/scalosure/scalosure/collection/immutable/List.js
goog.provide('scalosure.collection.immutable.List');

/** @constructor */
scalosure.collection.immutable.List = function(underlying) {
	this.underlying = underlying;
};

scalosure.collection.immutable.List.prototype.foreach = function(fn) {

	var self = this;

	for(var i = 0; i < self.underlying.length; i++) {
		fn(self.underlying[i]);
	}
};

scalosure.collection.immutable.List.prototype.map = function(fn) {

	var self = this;
	var newarray = [];

	for(var i = 0; i < self.underlying.length; i++) {
		newarray.push(fn(self.underlying[i]));
	}

	return new scalosure.collection.immutable.List(newarray);
};

scalosure.collection.immutable.List.prototype.isEmpty = function() {
	return (this.underlying.length == 0);
};

scalosure.collection.immutable.List.prototype.head = function() {
	return this.underlying.slice(0,1)[0];
};

scalosure.collection.immutable.List.prototype.tail = function() {
	return new scalosure.collection.immutable.List(this.underlying.slice(1));
};

scalosure.collection.immutable.List.prototype.exists = function(p) {

	var these = this;

	while (!these.isEmpty()) {
		if (p(these.head())) return true;
		these = these.tail();
	}

	return false;
};

scalosure.collection.immutable.List.prototype.headOption = function() {
	return (this.isEmpty()) ? new scalosure.None() : new scalosure.Some(this.head());
};

scalosure.collection.immutable.List.prototype.$apply = function(index) {
	return this.underlying[index];
};

scalosure.collection.immutable.List.prototype.filter = function(fn) {

	var self = this;
	var newarray = [];

	for(var i = 0; i < self.underlying.length; i++) {
		if(fn(self.underlying[i])) { newarray.push(self.underlying[i]); }
	}

	return new scalosure.collection.immutable.List(newarray);
};

scalosure.collection.immutable.List.$apply = function(xs) {
	return new scalosure.collection.immutable.List(xs);
};

File: library/src/main/scala/scalosure/collection/mutable/List.scala
package scalosure.collection.mutable

import scalosure._

class List(xs:JsArray[Any]) {

  def foreach(fn:(Any) => Unit):Unit = {
    var i = 0
    while(i < xs.length) {
      fn(xs(i))
      i = i + 1
    }
  }

  def map(fn:(Any) => Any):List = {
      var rst = JsArray.empty[Any]
      foreach { x => rst.push(fn(x)) }
      new List(rst)
  }

  def filter(fn:(Any) => Boolean):List = {
      var rst = JsArray.empty[Any]
      foreach { x => if(fn(x)) rst.push(x) }
      new List(rst)
  }

  def exists(fn:(Any) => Boolean):Boolean = {
      var rst = false
      foreach { x => if(fn(x)) { rst = true } }
      rst
  }

  def append(x:Any):Unit = {
    xs.push(x)
  }

  def toArray():JsArray[Any] = xs
}

